File: frontend/types/borrowedBook.ts
// types/borrowedBook.ts
import { BookRead } from './scanTypes';
import { Contact } from './contact';

export enum BorrowStatus {
	ACTIVE = 'active',
	RETURNED = 'returned',
	OVERDUE = 'overdue'
}

export interface BorrowedBook {
	id: number;
	book_id: number;
	user_id: number;
	contact_id?: number;
	contact?: Contact;
	borrowed_from: string; // Legacy, gardé pour rétrocompatibilité
	borrowed_date: string;
	expected_return_date?: string;
	return_date?: string;
	status: BorrowStatus;
	notes?: string;
	created_at: string;
	updated_at: string;
	calendar_event_id?: string;

	// Relations
	book?: BookRead;
}

export interface BorrowedBookCreate {
	book_id: number;
	contact: number | string | {
		name: string;
		email?: string;
		phone?: string;
		notes?: string;
	};
	borrowed_date?: string;
	expected_return_date?: string;
	notes?: string;
}

export interface BorrowedBookUpdate {
	contact?: number | string | {
		name: string;
		email?: string;
		phone?: string;
		notes?: string;
	};
	borrowed_date?: string;
	expected_return_date?: string;
	notes?: string;
	calendar_event_id?: string;
}

export interface BorrowStatistics {
	total: number;
	active: number;
	returned: number;
	overdue: number;
}

File: backend/app/schemas/BorrowedBook.py
from typing import Optional, Dict, Any
from sqlmodel import SQLModel
from datetime import datetime
from app.models.BorrowedBook import BorrowStatus
from app.schemas.Contact import ContactRead


class BorrowedBookRead(SQLModel):
    """Schema for reading a borrowed book record"""
    id: int
    book_id: int
    book: Optional["BookRead"] = None
    contact_id: Optional[int] = None
    contact: Optional[ContactRead] = None
    borrowed_from: str  # Legacy, gardé pour rétrocompatibilité
    borrowed_date: datetime
    expected_return_date: Optional[datetime] = None
    actual_return_date: Optional[datetime] = None
    status: BorrowStatus
    notes: Optional[str] = None
    calendar_event_id: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class BorrowedBookCreate(SQLModel):
    """Schema for creating a borrowed book record.

    Le champ contact accepte plusieurs formats :
    - int : ID d'un contact existant
    - str : Nom d'un contact (sera créé s'il n'existe pas)
    - dict : Objet avec 'name' et optionnellement 'email', 'phone', 'notes'
    """
    book_id: int
    contact: int | str | Dict[str, Any]
    borrowed_date: Optional[datetime] = None  # Defaults to now
    expected_return_date: Optional[datetime] = None
    notes: Optional[str] = None


class BorrowedBookUpdate(SQLModel):
    """Schema for updating a borrowed book record"""
    contact: Optional[int | str | Dict[str, Any]] = None
    borrowed_date: Optional[datetime] = None
    expected_return_date: Optional[datetime] = None
    actual_return_date: Optional[datetime] = None
    status: Optional[BorrowStatus] = None
    notes: Optional[str] = None
    calendar_event_id: Optional[str] = None


class BorrowedBookReturn(SQLModel):
    """Simplified schema for marking a borrowed book as returned"""
    actual_return_date: Optional[datetime] = None  # Defaults to now


class BorrowedBookStats(SQLModel):
    """Statistics for borrowed books"""
    total_borrowed: int
    currently_borrowed: int
    overdue: int
    returned: int


# Import BookRead at the end to avoid circular imports
from app.schemas.Book import BookRead
BorrowedBookRead.model_rebuild()
